[popover] Implement toggle event task

nt1m opened the pull request:
#### 53fc7cf368d2b375ec52fe5c8b33492b785340a8
<pre>
[popover] Implement toggle event task
<a href="https://bugs.webkit.org/show_bug.cgi?id=252215">https://bugs.webkit.org/show_bug.cgi?id=252215</a>
rdar://105703645

Reviewed by Darin Adler.

<a href="https://html.spec.whatwg.org/multipage/popover.html#queue-a-popover-toggle-event-task">https://html.spec.whatwg.org/multipage/popover.html#queue-a-popover-toggle-event-task</a>

Called in hidePopover/showPopover. In order to ensure only one event is emitted, we store the last queued oldState/newState,
then the queued task checks if the newState up to date and that there is still a queued task. The oldest &quot;oldState&quot; is always
preserved since we&apos;re collapsing multiple toggle events in one.

Also do other fixes to make html/semantics/popovers/popover-events.html pass:
- Make onbeforetoggle attribute actually work by adding it to IDL and attribute names
- Hide popover without firing events if element is removed from tree

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-document-open-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-removal-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-events.html:
* Source/WebCore/dom/GlobalEventHandlers.idl:
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::queuedToggleEventData):
(WebCore::PopoverData::setQueuedToggleEventData):
(WebCore::PopoverData::clearQueuedToggleEventData):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::removedFromAncestor):
(WebCore::HTMLElement::queuePopoverToggleEventTask):
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/HTMLElement.h:

Canonical link: <a href="https://commits.webkit.org/260728@main">https://commits.webkit.org/260728@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0aa6ad5a5d1dd3e8aca55214f2c268d5f1dff7be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109182 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18262 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/709 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118417 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113064 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19721 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9564 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101423 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114939 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98014 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42958 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29668 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84692 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11039 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31011 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11766 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17134 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50614 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7399 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13386 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/PopoverData.h:
```diff
@@ -49,9 +54,14 @@ class PopoverData {
     Element* previouslyFocusedElement() const { return m_previouslyFocusedElement.get(); }
     void setPreviouslyFocusedElement(Element* element) { m_previouslyFocusedElement = element; }
 
+    std::optional<PopoverToggleEventData> queuedToggleEventData() { return m_queuedToggleEventData; }
```
This should be `const`.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1233,6 +1236,25 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
+void HTMLElement::queuePopoverToggleEventTask(PopoverVisibilityState oldState, PopoverVisibilityState newState)
+{
+    if (auto queuedEventData = popoverData()->queuedToggleEventData())
+        oldState = (*queuedEventData).oldState;
```
Can use `->` here if you like.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1233,6 +1236,25 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
+        auto queuedEventData = popoverData()->queuedToggleEventData();
+        if (!queuedEventData)
+            return;
+        if ((*queuedEventData).newState != newState)
```
Can use `->` here if you like.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1233,6 +1236,25 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
+        if ((*queuedEventData).newState != newState)
+            return;
+        popoverData()->clearQueuedToggleEventData();
+        auto visibilityStateToString = [](PopoverVisibilityState state) {
```
Since this has narrow scope, it doesn’t need such a long name. Could just call it `string` even.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1233,6 +1236,25 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
+        auto visibilityStateToString = [](PopoverVisibilityState state) {
+            return state == PopoverVisibilityState::Hidden ? "closed"_s : "open"_s;
+        };
+        dispatchEvent(ToggleEvent::create(eventNames().toggleEvent, { EventInit { }, visibilityStateToString((*queuedEventData).oldState), visibilityStateToString((*queuedEventData).newState) }, Event::IsCancelable::No));
```
Can use `->` here if you like.

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -44,6 +44,7 @@ struct TextRecognitionResult;
 
 enum class PageIsEditable : bool;
 enum class EnterKeyHint : uint8_t;
+enum class PopoverVisibilityState : bool;
```
WebKit coding style suggests keeping these sorted alphabetically.

nt1m committed:
[popover] Implement toggle event task
https://bugs.webkit.org/show_bug.cgi?id=252215
rdar://105703645

Reviewed by Darin Adler.

https://html.spec.whatwg.org/multipage/popover.html#queue-a-popover-toggle-event-task

Called in hidePopover/showPopover. In order to ensure only one event is emitted, we store the last queued oldState/newState,
then the queued task checks if the newState up to date and that there is still a queued task. The oldest "oldState" is always
preserved since we're collapsing multiple toggle events in one.

Also do other fixes to make html/semantics/popovers/popover-events.html pass:
- Make onbeforetoggle attribute actually work by adding it to IDL and attribute names
- Hide popover without firing events if element is removed from tree

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-document-open-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-removal-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-events.html:
* Source/WebCore/dom/GlobalEventHandlers.idl:
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::queuedToggleEventData):
(WebCore::PopoverData::setQueuedToggleEventData):
(WebCore::PopoverData::clearQueuedToggleEventData):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::removedFromAncestor):
(WebCore::HTMLElement::queuePopoverToggleEventTask):
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/HTMLElement.h:

Canonical link: https://commits.webkit.org/260728@main
